core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala

```diff
@@ -0,0 +1,60 @@
+import org.broadinstitute.dsde.rawls.util.FutureSupport
+import spray.json.{JsObject, JsString}
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+object SnapshotService {
+
+  def constructor(dataSource: SlickDataSource, workspaceManagerDAO: WorkspaceManagerDAO, terraDataRepoUrl: String)(userInfo: UserInfo)
+                 (implicit executionContext: ExecutionContext): SnapshotService = {
+    new SnapshotService(userInfo, dataSource, workspaceManagerDAO, terraDataRepoUrl)
+  }
+
+}
+
+class SnapshotService(protected val userInfo: UserInfo, dataSource: SlickDataSource, workspaceManagerDAO: WorkspaceManagerDAO, terraDataRepoUrl: String)(implicit protected val executionContext: ExecutionContext) extends FutureSupport {
+
+  def CreateSnapshot(workspaceName: WorkspaceName, dataRepoSnapshot: DataRepoSnapshot): Future[DataRepoSnapshotReference] = createSnapshot(workspaceName, dataRepoSnapshot)
+  def GetSnapshot(workspaceName: WorkspaceName, snapshotId: String): Future[DataRepoSnapshotReference] = getSnapshot(workspaceName, snapshotId)
+
+  def createSnapshot(workspaceName: WorkspaceName, snapshot: DataRepoSnapshot): Future[DataRepoSnapshotReference] = {
+    for {
+      workspaceIdOpt <- dataSource.inTransaction { dataAccess => dataAccess.workspaceQuery.getWorkspaceId(workspaceName) }
+      workspaceId = workspaceIdOpt.getOrElse(throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"Workspace $workspaceName not found")))
+      stubExists <- workspaceStubExists(workspaceId, userInfo)
+      _ <- if(!stubExists) { workspaceManagerDAO.createWorkspace(workspaceId, userInfo) } else Future.successful()
+      dataRepoReference = JsObject.apply(("instance", JsString(terraDataRepoUrl)), ("snapshot", JsString(snapshot.snapshotId)))
+      res <- workspaceManagerDAO.createDataReference(workspaceId, snapshot.name, DataReferenceType.DataRepoSnapshot.toString, dataRepoReference, CloningInstructions.COPY_NOTHING.toString, userInfo)
+    } yield {
+      DataRepoSnapshotReference(res.getReferenceId.toString, res.getName, res.getWorkspaceId.toString, Option(res.getReferenceType.toString), Option(res.getReference), res.getCloningInstructions.toString)
+    }
+  }
+
+  def getSnapshot(workspaceName: WorkspaceName, snapshotId: String): Future[DataRepoSnapshotReference] = {
+    for {
+      workspaceIdOpt <- dataSource.inTransaction { dataAccess => dataAccess.workspaceQuery.getWorkspaceId(workspaceName) }
+      workspaceId = workspaceIdOpt.getOrElse(throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"Workspace $workspaceName not found")))
+      res <- workspaceManagerDAO.getDataReference(workspaceId, UUID.fromString(snapshotId), userInfo)
+    } yield {
+      DataRepoSnapshotReference(res.getReferenceId.toString, res.getName, res.getWorkspaceId.toString, Option(res.getReferenceType.toString), Option(res.getReference), res.getCloningInstructions.toString)
+    }
+  }
```

Comment:
For both create and get: I like deferring all authz to Workspace Manager, and keeping Rawls thin. However, it means that users can fish for existence/non-existence of workspaces that they don't have access to.

For create: this doesn't respect whether or not the workspace is locked ... which of course is only known by Rawls, not Sam or WSM.

You may need to copy/import WorkspaceService's getWorkspaceContextAndPermissions :( I suppose you could put these methods inside WorkspaceService, though I agree that's already too big.

Reply:
sorry if i'm being totally dense and missing something obvious, but i'm not seeing how deferring to WSM for authz is going to reveal extra information about existence of workspaces. if it does, shouldn't we ensure that WSM doesn't reveal that information?

Reply:
I'll sit with this and try to illustrate a use case - or admit I'm wrong and the code works as-is!

Reply:
okay, so consider two use cases:
1. You request a snapshot id in a workspace that does not exist
2. You request a snapshot id in a workspace that does exist, but you do not have permission to read

In the first case, line 46 of this PR will trigger, and the user will see a `RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, s"Workspace $workspaceName not found"))`

In the second case, we'll pass the request on to WSM, which authorizes against Sam. Sam will respond `false`, and WSM will catch that response and throw an error `User ${email} is not authorized to read workspace ${workspaceId} or it doesn't exist.`, (see https://github.com/DataBiosphere/terra-workspace-manager/blob/master/src/main/java/bio/terra/workspace/service/iam/SamService.java#L66), which will get propagated up.

A malicious user could look at the difference between these error messages to figure out if the workspace exists or  not.

Reply:
Ah! So they'd be able to infer because of the different messages from the different sources. That's a good catch, thanks for bearing with me on this :)

Reply:
yup exactly. And I think WSM does the right thing about not leaking that info already.